src/Optimization/hiopDualsUpdater.cpp

```diff
@@ -413,39 +413,63 @@ bool hiopDualsLsqUpdateLinsysAugSparse::do_lsq_update(hiopIterate& iter,
         
 #endif  // HIOP_USE_STRUMPACK
       }
-    } //KS: /end of CPU mode/ do not put CU SOLVER anywhere above this!!!!!
-else {
+      //KS: /end of CPU mode/ do not put CU SOLVER anywhere above this!!!!!
+    } else {  
       //
-      // we're on device
+      // We're on device
       //
-#ifdef HIOP_USE_CUSOLVER // This is our first choice on the device!
+      // Under hybrid compute_mode, LSQ-based initialization can be done using CPU sparse linear solvers.
+      // Under gpu compute_mode, which is work in progress, the initialization should be done only using
+      // GPU sparse linear solvers.
+      
+#ifdef HIOP_USE_CUSOLVER 
+      if(compute_mode == "gpu") {
         assert((linear_solver == "cusolver" || linear_solver == "auto") &&
                "the value for duals_init_linear_solver_sparse is invalid and should have been corrected during "
                "options processing");
-               hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
+      }
+
+      // This is our first choice on the device.
+      if(linear_solver == "cusolver" || linear_solver == "auto") {
+        hiopLinSolverIndefSparseCUSOLVER *p = new hiopLinSolverIndefSparseCUSOLVER(n, nnz, nlp_);
         nlp_->log->printf(hovSummary,
                           "LSQ Dual Initialization --- KKT_SPARSE_XDYcYd linsys: using CUSOLVER on device as an "
                           "indefinite solver, size %d (%d cons)\n",
                           n,
                           neq+nineq);
         p->setFakeInertia(neq + nineq);
         lin_sys_ = p;
-#endif 
-#if defined(HIOP_USE_STRUMPACK)
+      }
+#else // of #ifdef HIOP_USE_CUSOLVER 
+      //under compute mode gpu, at this point we don't have a sparse linear solver 
+      if(compute_mode == "gpu") {
+        if(linear_solver == "auto") {
+          assert("HiOp was not built with a sparse GPU and cannot fullfil the requirement of the option "
+                 "'duals_init_linear_solver_sparse'. Either build with a supported GPU sparse solver or "
+                 "change compute mode to hybrid, which will allow using a CPU sparse solver.");
+        } else {
+          assert("Impossible to deploy the (CPU?) sparse linear solver specified by option "
+                 "'duals_init_linear_solver_sparse' under gpu compute mode. Either build with a supported "
+                 "GPU sparse solver or change compute mode to hybrid, which will allow using CPU sparse solvers.");
```

Comment:
Two comments:
1. when we use this function, `do_lsq_update`, to initialize dual variables, we can return `false` whenever assertion fails. Then the outer loop ([here](https://github.com/LLNL/hiop/blob/e29104a18e380f9a537ddc2a960a4e34a95398cf/src/Optimization/hiopDualsUpdater.hpp#L139-L159)) will set duals to `zero`
2. this function can also be used in standard ipm iterations, e.g., solve the duals for quasi-newton method, and hence the error message may need to be modified. That is, the problem is not about `duals_init_linear_solver_sparse`, but `linear_solver_sparse` when `duals_update_type = lsq`.

Reply:
good point, will revisit